File: 2_CoinChange.cpp
#include <iostream>
#include <algorithm>

using namespace std;

class CoinChange
{
    int coins[100];
    int n, amount;

public:

    // Input function
    void getData()
    {
        cout << "Enter number of coin denominations: ";
        cin >> n;

        cout << "Enter coin denominations:\n";

        for(int i = 0; i < n; i++)
        {
            cin >> coins[i];
        }

        cout << "Enter amount to pay: ";
        cin >> amount;
    }

    // Sort denominations in descending order
    void sortCoins()
    {
        for(int i = 0; i < n - 1; i++)
        {
            for(int j = i + 1; j < n; j++)
            {
                if(coins[i] < coins[j])
                {
                    swap(coins[i], coins[j]);
                }
            }
        }
    }

    // Greedy Coin Change
    void coinChange()
    {
        sortCoins();

        int count = 0;

        cout << "\nCoins used:\n";

        for(int i = 0; i < n; i++)
        {
            while(amount >= coins[i])
            {
                amount = amount - coins[i];

                cout << coins[i] << " ";

                count++;
            }
        }

        cout << "\n\nMinimum number of coins = "
             << count;
    }
};

int main()
{
    CoinChange c;

    c.getData();

    c.coinChange();

    return 0;
}

File: 2_CoinChange.java
import java.util.Scanner;

public class CoinChange {
    private int[] coins = new int[100];
    private int n, amount;
    private Scanner sc = new Scanner(System.in);

    public void getData() {
        System.out.print("Enter number of coin denominations: ");
        n = sc.nextInt();

        System.out.println("Enter coin denominations:");
        for (int i = 0; i < n; i++) {
            coins[i] = sc.nextInt();
        }

        System.out.print("Enter amount to pay: ");
        amount = sc.nextInt();
    }

    public void sortCoins() {
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (coins[i] < coins[j]) {
                    int temp = coins[i];
                    coins[i] = coins[j];
                    coins[j] = temp;
                }
            }
        }
    }

    public void coinChange() {
        sortCoins();

        int count = 0;
        System.out.println("\nCoins used:");

        for (int i = 0; i < n; i++) {
            while (amount >= coins[i]) {
                amount = amount - coins[i];
                System.out.print(coins[i] + " ");
                count++;
            }
        }

        System.out.println("\n\nMinimum number of coins = " + count);
    }

    public static void main(String[] args) {
        CoinChange c = new CoinChange();
        c.getData();
        c.coinChange();
    }
}
